src/particles/incflo_PCEvolve.cpp

```diff
@@ -0,0 +1,181 @@
+                                                                  &((*a_vmac)[grid]),
+                                                                  &((*a_wmac)[grid])) };
+
+            //array of these pointers to pass to the GPU
+            GpuArray<Array4<const Real>, AMREX_SPACEDIM>
+                const umacarr {{AMREX_D_DECL((*fab[0]).array(),
+                                             (*fab[1]).array(),
+                                             (*fab[2]).array() )}};
+
+            ParallelFor(n, [=] AMREX_GPU_DEVICE (int i)
+            {
+                ParticleType& p = p_pbox[i];
+                if (p.id() <= 0) { return; }
+
+                ParticleReal v[AMREX_SPACEDIM];
+                mac_interpolate(p, plo, dxi, umacarr, v);
+
+                if (ipass == 0) {
+                    for (int dim=0; dim < AMREX_SPACEDIM; dim++)
+                    {
+                        v_ptr[dim][i] = p.pos(dim);
+                        p.pos(dim) += static_cast<ParticleReal>(ParticleReal(0.5)*a_dt*v[dim]);
+
+                        // Reflect off bottom boundary
+                        if (dim == 2 && !is_periodic_in_z && p.pos(dim) < plo[2])
+                        {
+                            p.pos(dim) = 2.0*plo[dim] - p.pos(dim);
+                            v_ptr[dim][i] *= -1.0;
+                        }
+                        // Reflect off top boundary
+                        if (dim == 2 && !is_periodic_in_z && p.pos(dim) > phi[2])
+                        {
+                            p.pos(dim) = 2.0*phi[dim] - p.pos(dim);
+                            v_ptr[dim][i] *= -1.0;
+                        }
+                    }
+                } else {
+                    for (int dim=0; dim < AMREX_SPACEDIM; dim++)
+                    {
+                        p.pos(dim) = v_ptr[dim][i] + static_cast<ParticleReal>(a_dt*v[dim]);
+                        v_ptr[dim][i] = v[dim];
+
+                        // Reflect off bottom boundary
+                        if (dim == 2 && !is_periodic_in_z && p.pos(dim) < plo[2])
+                        {
+                            p.pos(dim) = 2.0*plo[dim] - p.pos(dim);
+                            v_ptr[dim][i] *= -1.0;
+                        }
+                        // Reflect off top boundary
+                        if (dim == 2 && !is_periodic_in_z && p.pos(dim) > phi[2])
+                        {
+                            p.pos(dim)     = 2.0*phi[dim] - p.pos(dim);
+                            v_ptr[dim][i] *= -1.0;
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    ParmParse pp("cylinder");
+
+    Real cyl_radius;
+    pp.get("radius",cyl_radius);
+
+    Array<Real,3> cyl_center;
+    pp.get("center",cyl_center);
+    Real x_ctr = cyl_center[0];
+    Real y_ctr = cyl_center[1];
+    Real z_ctr = cyl_center[2];
+
+    // Remove particles that are outside of the cylindner
```

Comment:
It seems like maybe this part where particles outside a cylinder are removed is specific to a particular problem? Maybe a switch would be appropriate here? But then you'd probably want to also supply addition bcs for the particles (since now it seems only top and bottom reflect bcs exist)? ... Just trying to think about gotchas if another user wants particles, but perhaps a comment or warning is good enough for now.